api/v3/PaymentProcessor.php

```diff
@@ -124,11 +124,22 @@ function _civicrm_api3_payment_processor_getlist_defaults(&$request) {
  *   API result array.
  *
  * @throws \API_Exception
+ * @throws \CiviCRM_API3_Exception
  */
 function civicrm_api3_payment_processor_pay($params) {
+  /* @var CRM_Core_Payment $processor */
   $processor = Civi\Payment\System::singleton()->getById($params['payment_processor_id']);
   $processor->setPaymentProcessor(civicrm_api3('PaymentProcessor', 'getsingle', ['id' => $params['payment_processor_id']]));
   try {
+    $processor->setContributionID($params['contribution_id']);
+    $processor->setInvoiceID($params['invoice_id'] ?? '');
```

Comment:
@eileenmcnaughton What did we agree in terms of invoice_id?  This means that we'll end up with an empty invoice ID if not passed in via params. Maybe that's ok!

Reply:
@mattwire yep - it has to be passed in! We talked about Order.create adding one here https://lab.civicrm.org/dev/financial/issues/78 but haven't resolved that & although I marked it 'high' here https://lab.civicrm.org/dev/financial/issues/76 I'm re-thinking that & am probably not seeing it as something to worry about until some of the other things have been done